File: lis.js
//Longest Increasing Subsequence

function LIS() {

       this.vec = []

       this.len = -1

       this.L = []

       this.output = []
}

LIS.prototype = {

       constructor: LIS,

       set: function( arr ) {

            this.vec = arr

            this.len = arr.length  

            for(var i = 0; i < this.len; i++) this.L[ i ] = 0

       },

       solve: function() {
      
            this.dynamic() 

            this.getSubseq() 

          return this            
       },  

       dynamic: function() {

            var n = this.len - 1, 

                max = 0;

                this.L[ n ] = 1
                
                for(var i = n - 1; i >= 0; i--) {

                        max = 0

                        for(var j = i + 1; j <= n; j++) {

                                if(this.vec[ j ] >= this.vec[ i ] && this.L[ j ] > max) {

                                   max = this.L[ j ] 
                                }
                        } 

                        this.L[ i ] = 1 + max 
                }
       },

       getSubseq: function() {

            var max = this.L[ 0 ], 

                posMax = 0, 

                output = [], 

                k = 0

                for(var i = 1; i < this.L.length; i++) {

                    if(this.L[ i ] > max) {

                       max = this.L[ i ]

                       posMax = i
                    } 
                }
                                
            output[ k++ ] = this.vec[ posMax ]

            max--

            for(var j = posMax + 1; j <= this.L.length; j++) {

                    if(this.vec[ j ] > this.vec[ posMax ] && this.L[ j ] == max) {

                          output[ k++ ] = this.vec[ j ]

                          max--
                    }
                    
            } 

            this.output = output
       },

       toString: function() {

                return this.output.toString()
       },

       toArray: function() {

                return this.output
       },

       getLength: function() {

                return this.output.length; 
       }        
}

File: lis.c
/**
 *  Longest Incresing Subsequence
 *
 */
#include <stdio.h>

int i, 
    j, 
    n,    
    max, 
    v[20] = {4,1,7,6,7},
    L[20];  

FILE *fin,*fout;

void read();
void readFromFile(const char *);
void solve_dynamic_programming();
void display();

int main() {

    //read the data
    //read();

    readFromFile("lis.in");

    //solve the problem with dynamic programming
    solve_dynamic_programming();    
  
    //display the results
    display();

    return(0);
}

void read() {

    printf("n=");
    scanf("%d",&n); 

    for(i=0;i<n;i++) {

        printf("v[%d]=",i); 
        scanf("%d",&v[i]);          
    }
}

//if you want we can read from File
void readFromFile(const char *filename) {

    fin = fopen(filename, "r");

    fscanf(fin, "%d", &n); 
     
    for( i = 0; i < n; i++) {

        fscanf(fin,"%d", &v[ i ]);
    }

    fclose( fin );
}


void solve_dynamic_programming() {

    L[ n - 1 ] = 1;

    for(i = n - 2; i >= 0; i--) {

        max = 0; 

        for(j = i + 1; j <= n - 1; j++) {

            if(v[ i ] <= v[ j ] && L[ j ] > max) {

               max = L[ j ];
            }
        }
            L[ i ] = 1 + max;
    }  
}

void display() {

    int k;
    //first we start with first slot
    max = L[0];

    fout = fopen("lis.out","w");

    //pos of the maximum
    int pos = 0;

    for(i = 1; i < n; i++) {

        if(L[ i ] > max) { 

           max = L[ i ];

           pos = i;
        } 
    }   

    printf("Longest Increasing SubSequence Length -> %d", max);

    fprintf(fout,"Longest Increasing Subsequence %d", max);

    printf("\n%d ", v[ pos ]);

    fprintf(fout,"\n%d ", v[ pos ]);

    max--;
    
    for(k = pos + 1; k < n; k++) {

        if(L[ k ] == max && v[k] >= v[ pos ]) {

           printf("%d ",v[ k ]);

           fprintf(fout, "%d ", v[ k ]);

           max--;
        }
    }    

    fclose( fout );
}